File: python/day7.py
import math


def check_target(nums: list[int], index: int, current_sum: int, target: int) -> bool:
    if index == len(nums):
        return current_sum == target

    if current_sum > target:
        return False

    if check_target(nums, index + 1, max(1, current_sum) * nums[index], target):
        return True

    return check_target(nums, index + 1, current_sum + nums[index], target)


def check_target_b(nums: list[int], index: int, current_sum: int, target: int) -> bool:
    if index == len(nums):
        return current_sum == target

    if current_sum > target:
        return False

    if index > 0 and check_target_b(
        nums,
        index + 1,
        current_sum * 10 ** (int(math.log10(abs(nums[index]))) + 1) + nums[index],
        target,
    ):
        return True

    if check_target_b(nums, index + 1, max(1, current_sum) * nums[index], target):
        return True

    return check_target_b(nums, index + 1, current_sum + nums[index], target)


def day7a() -> None:
    with open("../inputs/day7.txt", "r") as file:
        rows = []
        for line in file.readlines():
            parts = line.split(": ")
            target = int(parts[0])
            nums_str = parts[1].split(" ")
            nums = [int(ns) for ns in nums_str]
            rows.append((target, nums))

    sum_good = 0
    for row in rows:
        target, nums = row
        if check_target(nums, 0, 0, target):
            sum_good += target

    print(sum_good)


def day7b() -> None:
    with open("../inputs/day7.txt", "r") as file:
        rows = []
        for line in file.readlines():
            parts = line.split(": ")
            target = int(parts[0])
            nums_str = parts[1].split(" ")
            nums = [int(ns) for ns in nums_str]
            rows.append((target, nums))

    sum_good = 0
    for row in rows:
        target, nums = row
        if check_target_b(nums, 0, 0, target):
            sum_good += target

    print(sum_good)


day7b()

File: go/day7/day7.go
package day7

import (
	"fmt"
	"math"
	"strconv"
	"strings"

	"github.com/tadejsv/aoc-2024/utils"
)

type Row struct {
	target int64
	nums   []int64
}

func CheckTarget(nums []int64, index int, currentSum int64, target int64) bool {
	if index == len(nums) {
		return currentSum == target
	}

	if currentSum > target {
		return false
	}

	if CheckTarget(nums, index+1, currentSum+nums[index], target) {
		return true
	}

	return CheckTarget(nums, index+1, max(currentSum, 0)*nums[index], target)
}

func CheckTargetB(nums []int64, index int, currentSum int64, target int64) bool {
	if index == len(nums) {
		return currentSum == target
	}

	if currentSum > target {
		return false
	}

	numDigits := int(math.Floor(math.Log10(float64(nums[index])))) + 1
	if CheckTargetB(nums, index+1, currentSum*int64(math.Pow10(numDigits))+nums[index], target) {
		return true
	}

	if CheckTargetB(nums, index+1, currentSum+nums[index], target) {
		return true
	}

	return CheckTargetB(nums, index+1, max(currentSum, 0)*nums[index], target)
}

func Day7a() {
	lines, err := utils.ReadLines("../inputs/day7.txt")
	if err != nil {
		panic("Could not read lines")
	}

	rows := make([]Row, len(lines))
	for i, line := range lines {
		parts := strings.Split(line, ": ")
		target, _ := strconv.Atoi(parts[0])
		numsStr := strings.Split(parts[1], " ")
		nums := make([]int64, len(numsStr))
		for j, numStr := range numsStr {
			num, _ := strconv.Atoi(numStr)
			nums[j] = int64(num)
		}
		row := Row{target: int64(target), nums: nums}
		rows[i] = row
	}

	sumGood := int64(0)
	for _, row := range rows {
		if CheckTarget(row.nums, 1, row.nums[0], row.target) {
			sumGood += row.target
		}
	}
	fmt.Printf("%v\n", sumGood)
}

func Day7b() {
	lines, err := utils.ReadLines("../inputs/day7.txt")
	if err != nil {
		panic("Could not read lines")
	}

	rows := make([]Row, len(lines))
	for i, line := range lines {
		parts := strings.Split(line, ": ")
		target, _ := strconv.Atoi(parts[0])
		numsStr := strings.Split(parts[1], " ")
		nums := make([]int64, len(numsStr))
		for j, numStr := range numsStr {
			num, _ := strconv.Atoi(numStr)
			nums[j] = int64(num)
		}
		row := Row{target: int64(target), nums: nums}
		rows[i] = row
	}

	sumGood := int64(0)
	for _, row := range rows {
		if CheckTargetB(row.nums, 1, row.nums[0], row.target) {
			sumGood += row.target
		}
	}
	fmt.Printf("%v\n", sumGood)
}
